scripts/merge-script/strategies/api-merge.ts

```diff
@@ -0,0 +1,191 @@
+const COMMIT_HEADER_SEPARATOR = '\n\n';
+
+/**
+ * Merge strategy that primarily leverages the Github API. The strategy merges a given
+ * pull request into a target branch using the API. This ensures that Github displays
+ * the pull request as merged. The merged commits are then cherry-picked into the remaining
+ * target branches using the local Git instance. The benefit is that the Github merged state
+ * is properly set, but a notable downside is that PRs cannot use fixup or squash commits.
+ */
+export class GithubApiMergeStrategy extends MergeStrategy {
+  async merge(pullRequest: PullRequest): Promise<PullRequestFailure|null> {
+    const {githubTargetBranch, prNumber, targetBranches, requiredBaseSha, needsCommitMessageFixup} =
+        pullRequest;
+    // If the pull request does not have its base branch set to any determined target
+    // branch, we cannot merge using the API.
+    if (targetBranches.every(t => t !== githubTargetBranch)) {
+      return {status: PullRequestStatus.MISMATCHING_TARGET_BRANCH, context: targetBranches};
+    }
+
+    // In case a required base is specified for this pull request, check if the pull
+    // request contains the given commit. If not, return a pull request failure. This
+    // check is useful for enforcing that PRs are rebased on top of a given commit. e.g.
+    // a commit that changes the codeowner ship validation. PRs which are not rebased
+    // could bypass new codeowner ship rules.
+    if (requiredBaseSha && !this.git.hasCommit(TEMP_PR_HEAD_BRANCH, requiredBaseSha)) {
+      return {status: PullRequestStatus.UNSATISFIED_BASE_SHA, context: requiredBaseSha};
+    }
+
+    const method = this._getMergeActionFromPullRequest(pullRequest);
+    const cherryPickTargetBranches = targetBranches.filter(b => b !== githubTargetBranch);
+
+    // First cherry-pick the PR into all local target branches in dry-run mode. This is
+    // purely for testing so that we can figure out whether the PR can be cherry-picked
+    // into the other target branches. We don't want to merge the PR through the API, and
+    // then run into cherry-pick conflicts after the initial merge already completed.
+    const failure = await this._checkMergability(pullRequest, cherryPickTargetBranches);
+
+    // If the PR could not be cherry-picked into all  locally, then we know it can't be done
+    // through the Github API either. We abort merging and pass-through the failure.
+    if (failure !== null) {
+      return failure;
+    }
+
+    const mergeOptions: PullsMergeParams = {
+      pull_number: prNumber,
+      merge_method: method,
+      ...this.git.repoParams,
+    };
+
+    // If a commit message fixup is needed and we are merging using squash, we prompt
+    // for commit message changes. Note that this does not work with other merge methods
+    // as the Github API only allows commit message modifications for squash merging.
+    if (needsCommitMessageFixup && method === 'squash') {
```

Comment:
Should we error in cases in which something is marked as `needsCommitMessageFixup` but is a `method` that is not `squash`?

Reply:
I like that.

Reply:
done.